Amplify/Categories/API/Operation/RetryableGraphQLOperation.swift

```diff
@@ -0,0 +1,172 @@
+
+    var resultListener: OperationResultListener { get }
+
+    init(requestFactory: @escaping RequestFactory,
+         maxRetries: Int,
+         resultListener: @escaping OperationResultListener,
+         _ operationFactory: @escaping OperationFactory)
+
+    func start(request: GraphQLRequest<Payload>)
+
+    func shouldRetry(error: APIError?) -> Bool
+}
+
+// MARK: RetryableGraphQLOperationBehavior + default implementation
+extension RetryableGraphQLOperationBehavior {
+    public func start(request: GraphQLRequest<Payload>) {
+        attempts += 1
+        log.debug("[\(id)] - Try [\(attempts)/\(maxRetries)]")
+        let wrappedResultListener: OperationResultListener = { result in
+            if case let .failure(error) = result, self.shouldRetry(error: error as? APIError) {
+                self.log.debug("\(error)")
+                self.start(request: self.requestFactory())
+                return
+            }
+
+            if case let .failure(error) = result {
+                self.log.debug("\(error)")
+                self.log.debug("[\(self.id)] - Failed")
+            }
+
+            if case .success = result {
+                self.log.debug("[Operation \(self.id)] - Success")
+            }
+            self.resultListener(result)
+        }
+        underlyingOperation = operationFactory(request, wrappedResultListener)
+    }
+}
+
+// MARK: - RetryableGraphQLOperation
+public final class RetryableGraphQLOperation<Payload: Decodable>: Operation, RetryableGraphQLOperationBehavior {
+    public typealias Payload = Payload
+    public typealias OperationType = GraphQLOperation<Payload>
+
+    public var id: UUID
+    public var maxRetries: Int
+    public var attempts: Int = 0
+    public var requestFactory: RequestFactory
+    public var underlyingOperation: GraphQLOperation<Payload>?
+    public var resultListener: OperationResultListener
+    public var operationFactory: OperationFactory
+
+    public init(requestFactory: @escaping () -> GraphQLRequest<Payload>,
+                maxRetries: Int,
+                resultListener: @escaping OperationResultListener,
+                _ operationFactory: @escaping OperationFactory) {
+        self.id = UUID()
+        self.maxRetries = max(1, maxRetries)
+        self.requestFactory = requestFactory
+        self.operationFactory = operationFactory
+        self.resultListener = resultListener
+    }
+    public override func main() {
+        start(request: requestFactory())
+    }
+
+    public override func cancel() {
+        underlyingOperation?.cancel()
+    }
+
+    public func shouldRetry(error: APIError?) -> Bool {
+        guard case let .operationError(_, _, underlyingError) = error,
+              let authError = underlyingError as? AuthError else {
+                  return false
+              }
+
+        switch authError {
+        case .signedOut, .notAuthorized:
+            return attempts < maxRetries
+        default:
+            return false
+        }
```

Comment:
If needed, in the future we can have the retry criteria injected by the consumer